sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala

```diff
@@ -165,6 +164,17 @@ trait FileFormat {
     }
   }
 
+  /**
+   * Create a file metadata struct column containing fields supported by the given file format.
+   */
+  def createFileMetadataCol(): AttributeReference = {
```

Comment:
do we expect implementations to override it?

Reply:
If we don't need this API now, let's remove it and put this function in `object FileFormat`. It's much harder to remove/change an API than adding an API later.

Reply:
This PR intentionally moved the method _out_ of `FileFormat`, as part of the effort to eliminate the the central code bottleneck antipattern of dumping everything in the `FileFormat` companion object -- which currently makes it impossible for file formats to customize any aspect of metadata handling. Absent a specific reason to forbid file formats from customizing this behavior, I'd prefer to allow it, so we don't have to keep coming back and changing the code.

Reply:
Also -- I had never seen evidence elsewhere in spark code, that we mark public methods `final` (or move them artificially to a companion object) solely to prevent the possibility that somebody might choose to override them in the future? Why the sudden requirement to do so here?